test/Rings/groebner.jl

```diff
@@ -147,7 +147,7 @@ end
 end
 
 @testset "f4" begin
-  R, (x1,x2,x3,x4) = polynomial_ring(GF(next_prime(2^28)), ["x1", "x2", "x3", "x4"], ordering=:degrevlex)
+  R, (x1,x2,x3,x4) = polynomial_ring(GF(next_prime(2^28)), ["x1", "x2", "x3", "x4"], internal_ordering=:degrevlex)
```

Comment:
I'm not sure what you mean. The result should be a degrevlex GB for the input ideal.

Reply:
Or is the difference only the order in which the terms are printed?

Reply:
No, I actually get different polynomials in the Gröbner basis with `internal_ordering = :lex` (the default). They still form a degrevlex GB as far as I can tell, but the test checks for the exact same polynomials and hence fails.

Reply:
Can I somehow see what you get as a result. If I use the current master of Oscar (w/o the required PRs for AA, Hecke, etc. mentioned above, I get the same reduced degrevlex GB computed by `groebner_basis_f4()` in the corresponding test case for any of these ring settings:

```
R, (x1,x2,x3,x4) = polynomial_ring(GF(next_prime(2^28)), ["x1", "x2", "x3", "x4"], ordering=:degrevlex)
R, (x1,x2,x3,x4) = polynomial_ring(GF(next_prime(2^28)), ["x1", "x2", "x3", "x4"], ordering=:lex)
R, (x1,x2,x3,x4) = polynomial_ring(GF(next_prime(2^28)), ["x1", "x2", "x3", "x4"])
```

Could you please paste the resulting GB on your side, so that I might be able to understand the problem.

Reply:
This is very strange. I get the following on OSCAR#master, so independent of this pull request:
```
julia> R, (x1,x2,x3,x4) = polynomial_ring(GF(next_prime(2^28)), ["x1", "x2", "x3", "x4"]);

julia> I = ideal(R,[x1+2*x2+2*x3+2*x4-1,
               x1^2+2*x2^2+2*x3^2+2*x4^2-x1,
               2*x1*x2+2*x2*x3+2*x3*x4-x2,
               x2^2+2*x1*x3+2*x2*x4-x3]);

julia> H = groebner_basis_f4(I);

julia> G = [x1 + 2*x2 + 2*x3 + 2*x4 + 268435458
                     x3^2 + 2*x2*x4 + 76695850*x3*x4 + 115043772*x4^2 + 115043768*x2 + 191739613*x3 + 230087535*x4
                     x2*x3 + 268435457*x2*x4 + 230087533*x3*x4 + 76695842*x4^2 + 210913575*x2 + 38347923*x3 + 153391692*x4
                     x2^2 + 2*x2*x4 + 153391692*x3*x4 + 230087538*x4^2 + 230087536*x2 + 115043768*x3 + 191739613*x4
                     x3*x4^2 + 238609298*x4^3 + 14913081*x2*x4 + 56338306*x3*x4 + 129246702*x4^2 + 263464432*x2 + 260150414*x3 + 258493405*x4
                     x2*x4^2 + 89478486*x4^3 + 29826162*x2*x4 + 263464432*x3*x4 + 238609297*x4^2 + 141674270*x2 + 9942054*x3
                     x4^4 + 35851649*x4^3 + 232885084*x2*x4 + 251179133*x3*x4 + 231479137*x4^2 + 191484965*x2 + 85955250*x3 + 167408122*x4];

julia> @test elements(H) == G
Test Failed at REPL[19]:1
  Expression: elements(H) == G
  ...
```
With `ordering = :degrevlex` the test passes.
The computed GB above is:
```
julia> H
Gröbner basis with elements
1 -> x1 + 2*x2 + 2*x3 + 2*x4 + 268435458
2 -> x3^2 + 2*x2*x4 + 76695850*x3*x4 + 115043772*x4^2 + 115043768*x2 + 191739613*x3 + 230087535*x4
3 -> x2*x3 + 268435457*x2*x4 + 230087533*x3*x4 + 76695842*x4^2 + 210913575*x2 + 38347923*x3 + 153391692*x4
4 -> x2^2 + 2*x2*x4 + 153391692*x3*x4 + 230087538*x4^2 + 230087536*x2 + 115043768*x3 + 191739613*x4
5 -> x3*x4^2 + 238609298*x4^3 + 14913081*x2*x4 + 56338306*x3*x4 + 129246702*x4^2 + 263464432*x2 + 260150414*x3 + 258493405*x4
6 -> x2*x4^2 + 89478486*x4^3 + 29826162*x2*x4 + 263464432*x3*x4 + 238609297*x4^2 + 141674270*x2 + 9942054*x3
7 -> x4^4 + 35851649*x4^3 + 232885084*x2*x4 + 251179133*x3*x4 + 231479137*x4^2 + 191484965*x2 + 85955250*x3 + 167408122*x4
with respect to the ordering
degrevlex([x1, x2, x3, x4])
```
And you are right: `H` contains exactly the same elements as `G` if one looks closely.
For example
```
julia> H[2]
x3^2 + 2*x2*x4 + 76695850*x3*x4 + 115043772*x4^2 + 115043768*x2 + 191739613*x3 + 230087535*x4

julia> G[2]
2*x2*x4 + 115043768*x2 + x3^2 + 76695850*x3*x4 + 191739613*x3 + 115043772*x4^2 + 230087535*x4
```
are indeed the same polynomials, the terms are just printed in a different order.
But now comes the really strange part:
```
julia> parent(G[2]) === parent(H[2])
true

julia> G[2] == H[2]
false

julia> G[2] - H[2]
2*x2*x4 + 115043768*x2 + 268435457*x2*x4 + 191739613*x3 + 153391691*x2 + 76695846*x3
```
Can you confirm this?

Versioninfo:
```
julia> Oscar.versioninfo(full = true)
OSCAR version 1.0.0-DEV - #master, 871f2244a5 -- 2024-02-15 13:44:43 +0100
  combining:
    AbstractAlgebra.jl   v0.37.6
    GAP.jl               v0.10.2
    Hecke.jl             v0.27.0
    Nemo.jl              v0.41.5
    Polymake.jl          v0.11.13
    Singular.jl          v0.22.2
  building on:
    Antic_jll               v0.201.500+0
    Arb_jll                 v200.2300.0+0
    Calcium_jll             v0.401.100+0
    FLINT_jll               v200.900.9+0
    GAP_jll                 v400.1200.200+7
    Singular_jll            v403.214.1400+0
    libpolymake_julia_jll   v0.11.2+0
    libsingular_julia_jll   v0.41.0+0
    polymake_jll            v400.1100.1+0
See `]st -m` for a full list of dependencies.

Julia Version 1.10.1
Commit 7790d6f0641 (2024-02-13 20:41 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 20 × 13th Gen Intel(R) Core(TM) i5-13600
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, goldmont)
Threads: 1 default, 0 interactive, 1 GC (on 20 virtual cores)
Environment:
  JULIA_EDITOR = vim
Official https://julialang.org/ release
```

Reply:
Yes, this I can confirm. The computed reduced GBs are always the same, but the ordering of the printing is different. So, for example, this is always `true`:
```
[g for g in terms(G[2], ordering=degrevlex(R))] == [h for h in terms(H[2], ordering=degrevlex(R))]
```
For `G[2]-H[2]` it is pretty clear what happens: The subtraction looks for the highest term in H[2] and keeps the first two terms of G[2] and then things mess up. I do not know how this subtraction is internally handled, but I assume it depends on the `internal_ordering` to be efficient.

Reply:
Ah, I understand. We did the polynomial construction that way because it is faster when the examples get bigger. I will make a new release for `AlgebraicSolving` with your fix.

Reply:
Thank you!